File: Projects/usaco/money.java
/*
ID: rombin82
LANG: JAVA
TASK: money
*/

import java.io.*;
import java.util.*;

public class money {
	public static void main(String[] args) throws IOException {
		Scanner in = new Scanner(new File("money.in"));
		PrintWriter stdout = new PrintWriter(new File("money.out"));
		Runtime.getRuntime().addShutdownHook(new Thread(() -> stdout.flush()));
		int V = in.nextInt();
		int N = in.nextInt();
		int[] coins = new int[V];
		for (int v = 0; v < V; v++)
			coins[v] = in.nextInt();
		in.close();
		int[] counts = new int[N + 1];
		counts[0] = 1;
		for (int v = 0; v < V; v++)
			for (int n = coins[v]; n <= N; n++)
				counts[n] += counts[n - coins[v]];
		stdout.println(counts[N]);
	}
}

File: Projects/usaco/money.kt
/*
ID: rombin82
LANG: JAVA
TASK: money
*/

import java.io.*
import java.util.*

object money {
    @Throws(IOException::class)
    fun main(args: Array<String>) {
        val stdin = Scanner(File("money.in"))
        val stdout = PrintWriter(File("money.out"))
        Runtime.getRuntime().addShutdownHook(Thread { stdout.flush() })
        val V = stdin.nextInt()
        val N = stdin.nextInt()
        val coins = IntArray(V)
        for (v in 0 until V)
            coins[v] = stdin.nextInt()
        stdin.close()
        val counts = IntArray(N + 1)
        counts[0] = 1
        for (v in 0 until V)
            for (n in coins[v]..N)
                counts[n] += counts[n - coins[v]]
        stdout.println(counts[N])
    }
}
